python/TestHarness/util.py

```diff
@@ -146,86 +154,94 @@ def formatResult(tester_data, result, options, color=True):
-        if not result:
-            formatted_result['s'] = (str(tester.getStatusMessage()), status.color)
-        else:
-            formatted_result['s'] = (result, status.color)
-
-    # Support the option of no name...
-    if 'n' in terminal_format:
-        formatted_result['n'] = (tester.getTestName(), None)
-
-    # Decorate and format timing
-    if options.timing and 't' in terminal_format:
-        actual = float(tester_data.getTiming())
-        int_len = len(str(int(actual)))
-        precision = min(3, max(0,(4-int_len)))
-        formatted_result['t'] = ('[' + '{0: <6}'.format('%0.*fs' % (precision, actual)) + ']', None)
-
-    # Special Case: When printing a pre-status _and_ a status, strip out the status
-    # if that status is identical to the pre-status. This is so we can support the
-    # printing of TestHarness statuses like RUNNING... while not printing two 'OK's
-    if (formatted_result['p'] and formatted_result['s']) \
-       and (formatted_result['p'][0].replace(' ', '') == formatted_result['s'][0].replace(' ', '')):
-        formatted_result['s'] = None
-
-    # Format the caveats if any
-    if tester.getCaveats() and 'c' in terminal_format:
+    formatted_results = {}
+    for format_key in terminal_format:
+        formatted_results[format_key] = None
+
+    # Populate formatted_results for those we support, with requested items
+    # specified by the user. Caveats and justifications are parsed outside of
+    # loop as these two items change based on character count consumed by others.
+    caveat_index = None
+    justification_index = None
+    for i, f_key in enumerate(terminal_format):
+        # Store the caveat request. We will use this later.
+        if str(f_key).lower() == 'c':
+            caveat_index = terminal_format[i]
+
+        # Store the justification request. We will use this later.
+        if str(f_key).lower() == 'j':
+            justification_index = terminal_format[i]
+
+        if str(f_key).lower() == 'p':
+            pre_result = ' '*(8-len(status.status)) + status.status
+            formatCase(f_key, (pre_result, status.color), formatted_results)
+
+        if str(f_key).lower() == 's':
+            if not result:
+                result = str(tester.getStatusMessage())
+
+            # refrain from printing a duplicate pre_result if it will match result
+            if 'p' in [x.lower() for x in terminal_format] and result == status.status:
+                formatCase(f_key, None, formatted_results)
+            else:
+                formatCase(f_key, (result, status.color), formatted_results)
+
+        if str(f_key).lower() == 'n':
+            formatCase(f_key, (tester.getTestName(), None), formatted_results)
+
+        if str(f_key).lower() == 't' and options.timing:
+            actual = float(tester_data.getTiming())
+            int_len = len(str(int(actual)))
+            precision = min(3, max(0,(4-int_len)))
```

Comment:
Could you add a comment what this does? I know because you told me but I probably won't remember in 6 months.